File: backend-haskell/test/Support/MockJwksServer.hs
-- | テスト専用の自プロセス内蔵JWKSサーバー(backend-java/backend-kotlin/backend-python/
-- backend-elixir/backend-c/backend-cppのモックJWKSサーバーと同じ役割)。実際にKeycloak/bffを
-- 起動せずにJwksVerifier/JwksCacheのRS256検証・kidキャッシュ・未知kid時の再取得ロジックを
-- 検証できる。`warp`(本番のREST実装と同じライブラリ)をそのままテスト用の使い捨てモックとしても
-- 再利用する
module MockJwksServer
  ( MockJwks
  , newMockJwks
  , addKey
  , resetKeys
  , mockJwksApp
  , withMockJwksServer
  ) where

import Data.Aeson (Value, object, (.=))
import Data.IORef (IORef, atomicModifyIORef', newIORef, readIORef)
import Data.Text (Text)
import Network.HTTP.Types (status200)
import Network.Wai (Application, responseLBS)
import Network.Wai.Handler.Warp (Port, testWithApplication)
import qualified Data.Aeson as Aeson

newtype MockJwks = MockJwks (IORef [Value])

newMockJwks :: IO MockJwks
newMockJwks = MockJwks <$> newIORef []

addKey :: MockJwks -> Text -> (Text, Text) -> IO ()
addKey (MockJwks ref) kid (n, e) =
  atomicModifyIORef' ref (\keys -> (keys ++ [keyJson kid n e], ()))

keyJson :: Text -> Text -> Text -> Value
keyJson kid n e = object ["kty" .= ("RSA" :: Text), "kid" .= kid, "use" .= ("sig" :: Text), "n" .= n, "e" .= e]

resetKeys :: MockJwks -> IO ()
resetKeys (MockJwks ref) = atomicModifyIORef' ref (const ([], ()))

mockJwksApp :: MockJwks -> Application
mockJwksApp (MockJwks ref) _req respond = do
  keys <- readIORef ref
  respond (responseLBS status200 [("Content-Type", "application/json")] (Aeson.encode (object ["keys" .= keys])))

-- | テストからは`around withMockJwksServer $ ... $ \(mock, port) -> ...`のように使う
-- (hspecの`around`は`ActionWith a -> IO ()`、すなわち1引数の継続を要求するため、タプルで渡す)。
-- portをJWKS URLの組み立てに使う
withMockJwksServer :: ((MockJwks, Port) -> IO a) -> IO a
withMockJwksServer action = do
  mock <- newMockJwks
  testWithApplication (pure (mockJwksApp mock)) (\port -> action (mock, port))

File: backend-elixir/test/support/mock_jwks_server.ex
defmodule BackendElixir.TestSupport.MockJwksServer.Plug do
  @moduledoc false
  use Plug.Router

  plug :match
  plug :dispatch

  get "/jwks" do
    keys = :persistent_term.get(:mock_jwks_keys, [])
    body = Jason.encode!(%{"keys" => keys})

    conn
    |> put_resp_content_type("application/json")
    |> send_resp(200, body)
  end
end

defmodule BackendElixir.TestSupport.MockJwksServer do
  @moduledoc """
  テスト専用の自プロセス内蔵JWKSサーバー(backend-java/backend-kotlin/backend-python/backend-c/
  backend-cppのモックJWKSサーバーと同じ役割)。実際にKeycloak/bffを起動せずにJwksVerifier/
  JwksCacheServerのRS256検証・kidキャッシュ・未知kid時の再取得ロジックを検証できる。
  本番実装(Plug.Router+Cowboy)と同じライブラリをそのままテスト用の使い捨てモックとしても
  再利用する(「本番はPlug/Cowboy、Verifierが叩く先はただのHTTPサーバーであれば何でもよい」
  という点を示す、backend-pythonが標準ライブラリのhttp.serverを使ったのと同じ考え方)
  """

  alias BackendElixir.TestSupport.MockJwksServer.Plug, as: MockPlug

  def start do
    ref = make_ref()
    {:ok, _pid} = Plug.Cowboy.http(MockPlug, [], ref: ref, port: 0)
    port = :ranch.get_port(ref)
    {:ok, %{ref: ref, port: port, jwks_url: "http://127.0.0.1:#{port}/jwks"}}
  end

  def stop(%{ref: ref}) do
    Plug.Cowboy.shutdown(ref)
  end

  @doc """
  :persistent_termでキー一覧を保持する(テスト実行プロセスとCowboyのリクエストハンドラは
  別プロセスのため、プロセス辞書ではなく全プロセス共有の状態として持つ、テスト専用の簡略化)
  """
  def add_key(kid, jwk_public_fields) do
    existing = :persistent_term.get(:mock_jwks_keys, [])
    key = Map.merge(%{"kty" => "RSA", "kid" => kid, "use" => "sig"}, jwk_public_fields)
    :persistent_term.put(:mock_jwks_keys, existing ++ [key])
  end

  def reset_keys do
    :persistent_term.put(:mock_jwks_keys, [])
  end
end
